main.rb

```diff
@@ -1,17 +1,140 @@
 module Enumerable
-    def my_each
-        for i in 0...self.size
-            yield(self[i])
-        end
-    self
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_each
+    i = 0
+    if block_given?
+      while i < length
+        yield(self[i])
+        i += 1
+      end
+      return self
+    end
+    to_enum
+  end
+
+  def my_each_with_index
+    i = 0
+    if block_given?
+      while i <= length - 1
+        yield(self[i], i)
+        i += 1
+      end
+      return self
     end
+    to_enum
+  end
+
+  def my_select
+    arr = []
+    return to_enum unless block_given?
+
+    my_each { |x| arr << x if yield(x) == true }
+    arr
+  end
 
-    def my_each_with_index
-        for i in 0...self.size
-            yield(self[i], i)
+  def my_all?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == false }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == false }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == false }
+    elsif arg.nil? == false
+      my_each { |x| return false if x != arg }
+    else
+      my_each { |x| return false if x.nil? == true }
```

Comment:
When I tried with a different case found a failing edge
```ruby
[1, false, 'hi', []].my_all?   #should return false but your current formation returns true.
```
Cause It checks for nil. When we check false.nil? it returns false but you are expecting it should be true.
![image](https://user-images.githubusercontent.com/38625956/79918496-51a05900-844e-11ea-8f75-00bac6562b0c.png)
So to cover that scenario I think the following change can do so.
```ruby
 my_each { |x| return false if x.nil? == true }
```
should be
 ```ruby
 my_each { |x| return false unless x } 
```